### ISSUE TITLE
Enable operators conv1d and conv2d

### ISSUE THREAD

**Issue by umacdevi** [torch-spyre/torch-spyre#444]
### What feature would you like to be added?

Run torch.conv1d and torch.conv2d on spyre

### Why is this feature needed?

The convolution operator is used by several models that are to be run on Spyre.

### Any alternative solution that you may want to share?

No

### Anything else we need to know?

No

**Comment by umacdevi** [comment]
depthwiseconv2d without padding works and has been tested with a few use cases. When padding is included, the following error occurs:

> what():  DtException: Program verification failed for core 0 node depthwiseconv2dnative
> Error message: Register initialization out of boundary:
> l3lu : EBR2 : -128

This is due to a bug in the backend for which a fix is expected to be pushed. There is also the workaround of not allocating input tensor1 at Segment 0 or at address 0. That will be tried if the backend fix takes time.

**Comment by umacdevi** [comment]
The following are the parameters for the model targeted for Q2:
```
description: 'site-packages/transformers/models/pixtral/modeling_pixtral.py:497 in forward, code: patch_embeds 
Input Tensor:
          shape: [1, 3, 392, 532]
Weight tensor:
          shape: [1024, 3, 14, 14]
Other params:
      - value: null  // bias
      - value: (14, 14)  // stride
      - value: (0, 0)  // padding
      - value: (1, 1)  // dilation
      - value: 1   // number of groups
```
For the above parameters, the current stack does a CPU fallback. Since large kernel sizes are not supported in the backend, the workaround is to perform a decomposition using bmm.

**Comment by umacdevi** [comment]
Decomposition using bmm is complete. However, though the tensor dimensions reaching bmm lowering are as expected, the tensor dimensions that are encoded in sdsc are incorrect.
E.g.,
For the convolution of input tensor [8, 3, 64, 64] with kernel tensor [8, 3, 3, 3], the resulting bmm is x[8, 8, 27] @ y[8, 27, 4096]. However, what's specified in the sdsc is [8, 64] @ [64, 4096]. The N (mb) dimension is dropped. This is currently under investigation.

**Comment by umacdevi** [comment]
Details of decomposition of  conv2d via unfold and bmm:

Let the  input image tensor be  (N, C_in, H_in, W_in) and filters be (C_out, C_in, K_h, K_w).

Steps:
1. Extract Patches (flattened submatrix fields): Use the torch.nn.Unfold module to slide a window (filter) over the input, flattening each sub-matrix of dimension K_h x K_w covered by the filter into a column. This reshapes the input to (N, C_in x K_h x K_w, L), where L is the number of total patches (output pixels).
2. Reshape Weights: Flatten the convolutional filters into a matrix of shape (C_out, C_in x K_h x K_w) using torch.reshape.

3. Batch Matrix Multiply (bmm): Since PyTorch's bmm applies to 3D tensors (B x M x K) and (B x K x N), add a batch dimension to the flattened weights (using torch.unsqueeze and torch.expand) and run torch.bmm. torch.bmm is run on expanded weight tensor of dimension (N, C_out, C_in x K_h x K_w)  and unfolded input tensor  (N, C_in x K_h x K_w, L). torch.bmm returns a tensor of dimension (N, C_out, L).

4. Reshape Output: Reorganize the result back to the 2D spatial feature map shape of C_out channels and N batches: (N, C_out, H_out, W_out). 

Reshaping weights using torch.reshape (Step 2) leads to mis-alignment in stick dimensions which seems hard to fix. Hence, implemented a custom reshape op for spyre, which makes the reshape fallback to CPU.

However, even a vanilla matmul with the dimensions of the transformed (unfolded and reshaped) matrices fails with dxp_standalone.

Some of the examples tried out are:

1. A = torch.randn(8, 64, 576, dtype=torch.float16)
  B = torch.randn(8, 576, 15876, dtype=torch.float16


2. A = torch.randn(1, 1024, 588, dtype=torch.float16)
  B = torch.randn(1, 588, 196701, dtype=torch.float16)


3. A = torch.randn(8, 64, 75, dtype=torch.float16)
  B = torch.randn(8, 75, 16384, dtype=torch.float16)

4. A = torch.randn(8, 8, 27, dtype=torch.float16)
   B = torch.randn(8, 27, 4096, dtype=torch.float16)


Torch.matmul(A,B) fails in each of the above cases.

The following is the DeepTools error:
``
INFO] [bundle] Generating /tmp/torchinductor_umamadev/tmpuas3f3aq/inductor-spyre/sdsc_fused_bmm_0_xb8fkzdl/bundle.mlir
terminate called after throwing an instance of 'DtException'
  what():  DtException: Program verification failed for core 0 node 0_identity
Error message: Register initialization out of boundary:
lxlu0 : LRF0 : -4294967296
``

**Comment by umacdevi** [comment]
The above matmuls are working in my build now. They were due to some local changes that led to sdsc differences.

**Comment by moriohara** [comment]
@umacdevi for issue tracking purposes, please have #2317 reviewed & merged first to close this issue, and open another issue to enable conv1d for granite-4-0-h, a Q3 target model. Thanks.

**Comment by umacdevi** [comment]
Sure @moriohara. Have just one review so far. Need 1 more to merge.

**Comment by moriohara** [comment]
@umacdevi please add a few test cases to the PR as @dgrove-oss suggested.
